walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,172 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.lang.Math;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.Map;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // Edge case if request's duration is over a day -> returns no options
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+        return Arrays.asList();
+    }
+    // Get all mandatory meeting attendees
+    Set<String> attendees = new HashSet<>(request.getAttendees()); 
+
+    // Get all optional attendees
+    Set<String> optAttendees = new HashSet<>(request.getOptionalAttendees());
+    
+    // Process which events are from mandatory or optional attendees
+    ArrayList<TimeRange> mandatoryEvents = new ArrayList<>();
+    ArrayList<TimeRange> optEvents =  new ArrayList<>();
+    HashMap<String, ArrayList<TimeRange>> optAttendeesEvents = new HashMap<>();
+
+    for (String i: optAttendees) {
+        optAttendeesEvents.put(i, new ArrayList<TimeRange>());
+    }
+
+    for (Event i: events) {
+        if (optAttendees.containsAll(i.getAttendees())) {
+            optEvents.add(i.getWhen());
+        } else if (attendees.containsAll(i.getAttendees())) {
+            mandatoryEvents.add(i.getWhen());
+        }
+    }
+
+    // request has only mandatory attendees so return time slots for them
+    if (attendees.isEmpty() && !(optAttendees.isEmpty())) {
+        return driver(optEvents, request);
+    } // request has no mandatory attendees so return time slots for optional attendees only
+    else if (!(attendees.isEmpty()) && optAttendees.isEmpty()) { 
+        return driver(mandatoryEvents, request);
```

Comment:
I think this is not quite correct for the case where there is no time where all optional attendees can meet.

Reply:
The comments were reversed for the 1st and 2nd case, sorry! The case you mentioned passes the unitTest "unavailableOptionalAttendees."

Reply:
`unavailableOptionalAttendees` is not the specific case I mentioned. The case I mentioned is if there is *no* free block for optional attendees. Then I expect the algorithm to suggest a time where the most optional attendees can meet, not no block.

So, for example (all attendees are optional):

```
A: ----------------     (unavailable whole day)
B:            -----
C  ------      
```

The current code as I read it will not offer no suggestion, but the better thing to do is to suggest the block where the most optional attendees can meet; namely the block in the middle of the day where B and C can meet.

Reply:
(I also think there isn't a test for this case.)